### PR TITLE
[MU3] Add more accidentals

### PR DESCRIPTION
**Pull request by Jojo-Schmitz**
see https://musescore.org/en/node/306457 and https://musescore.org/en/node/306456

* add missing Gould arrow down/up accidentals
* add more extended Helmholtz-Ellis accidentals
* add the most important (Spartan) Sagittal accidentals
* add curly braces to accidentals
* (almost) complete MusicXML im- and export

See also #6002 (Add support for triple sharp and triple flat, pending), #5737 (adding Wyschnegradsky accidentals, merged) and #6263 ([MusicXML] Add missing Gould arrow quartertone accidentals, merged)

### CODE REVIEW

**Comment by Ra-Yo-Su** [comment]
The 53 limit commas (Helmholts Ellis) are +/- 10.9.

Here's the data for the most important (Spartan) Sagittal accidentals:

Acc(AccidentalVal::NATURAL, 5.8, SymId::accSagittal5v7KleismaUp),
Acc(AccidentalVal::NATURAL, -5.8, SymId::accSagittal5v7KleismaDown),
Acc(AccidentalVal::NATURAL, 21.5, SymId::accSagittal5CommaUp),
Acc(AccidentalVal::NATURAL, -21.5, SymId::accSagittal5CommaDown),
Acc(AccidentalVal::NATURAL, 27.3, SymId::accSagittal7CommaUp),
Acc(AccidentalVal::NATURAL, -27.3, SymId::accSagittal7CommaDown),
Acc(AccidentalVal::NATURAL, 43.0, SymId::accSagittal25SmallDiesisUp),
Acc(AccidentalVal::NATURAL, -43.0, SymId::accSagittal25SmallDiesisDown),
Acc(AccidentalVal::NATURAL, 48.8, SymId::accSagittal35MediumDiesisUp),
Acc(AccidentalVal::NATURAL, -48.8, SymId::accSagittal35MediumDiesisDown),
Acc(AccidentalVal::NATURAL, 53.3, SymId::accSagittal11MediumDiesisUp),
Acc(AccidentalVal::NATURAL, -53.3, SymId::accSagittal11MediumDiesisDown),
Acc(AccidentalVal::NATURAL, 60.4, SymId::accSagittal11LargeDiesisUp),
Acc(AccidentalVal::NATURAL, -60.4, SymId::accSagittal11LargeDiesisDown),
Acc(AccidentalVal::NATURAL, 64.9, SymId::accSagittal35LargeDiesisUp),
Acc(AccidentalVal::NATURAL, -64.9, SymId::accSagittal35LargeDiesisDown),
Acc(AccidentalVal::NATURAL, 70.7, SymId::accSagittalSharp25SDown),
Acc(AccidentalVal::NATURAL, -70.7, SymId::accSagittalFlat25SUp),
Acc(AccidentalVal::NATURAL, 86.4, SymId::accSagittalSharp7CDown),
Acc(AccidentalVal::NATURAL, -86.4, SymId::accSagittalFlat7CUp),
Acc(AccidentalVal::NATURAL, 92.2, SymId::accSagittalSharp5CDown),
Acc(AccidentalVal::NATURAL, -92.2, SymId::accSagittalFlat5CUp),
Acc(AccidentalVal::NATURAL, 107.9, SymId::accSagittalSharp5v7kDown),
Acc(AccidentalVal::NATURAL, -107.9, SymId::accSagittalFlat5v7kUp),
Acc(AccidentalVal::NATURAL, 113.7, SymId::accSagittalSharp),
Acc(AccidentalVal::NATURAL, -113.7, SymId::accSagittalFlat),

File: libmscore/types.h [hunk 1]
```diff
@@ -285,6 +285,34 @@ enum class AccidentalType : char {
...
+      S17,
+      S18,
+      S19,
+      S20,
+      S21,
+      S22,
+      S23,
+      S24,
+      S25,
+      S26,
```
Comment (Jojo-Schmitz) [inline libmscore/types.h:314]: Need to come up with more useful names here I guess

File: libmscore/sym.cpp [hunk 2]
```diff
@@ -2683,10 +2683,10 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
...
       "4-string tab clef",
       "6-string tab clef",
-      "11 large diesis down, 3\u00b0 down [46 EDO]",
-      "11 large diesis up, (11L), (sharp less 11M), 3\u00b0 up [46 EDO]",
-      "11 medium diesis down, 1\u00b0[17 31] 2\u00b046 down, 1/4-tone down",
-      "11 medium diesis up, (11M), 1\u00b0[17 31] 2\u00b046 up, 1/4-tone up",
+      QT_TRANSLATE_NOOP("symUserNames", "11 large diesis down, 3\u00b0 down [46 EDO]"),
+      QT_TRANSLATE_NOOP("symUserNames", "11 large diesis up, (11L), (sharp less 11M), 3\u00b0 up [46 EDO]"),
+      QT_TRANSLATE_NOOP("symUserNames", "11 medium diesis down, 1\u00b0[17 31] 2\u00b046 down, 1/4-tone down"),
+      QT_TRANSLATE_NOOP("symUserNames", "11 medium diesis up, (11M), 1\u00b0[17 31] 2\u00b046 up, 1/4-tone up"),
```
Comment (Jojo-Schmitz) [inline libmscore/sym.cpp:2689]: the above 2 may fail to translate properly, due to the 2nd UTF char (for °) being followed by (hex) numbers

File: libmscore/sym.cpp [hunk 3]
```diff
@@ -2707,12 +2707,12 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
...
       "23 small diesis down",
       "23 small diesis up, (23S)",
-      "25 small diesis down, 2\u00b0 down [53 EDO]",
-      "25 small diesis up, (25S, ~5:13S, ~37S, 5C\u00a0plus\u00a05C), 2\u00b0 up [53 EDO]",
-      "35 large diesis down, 2\u00b0 down [50 EDO], 5/18-tone down",
-      "35 large diesis up, (35L, ~13L, ~125L,  sharp less 35M), 2\u00b050 up",
-      "35 medium diesis down, 1\u00b0[50] 2\u00b0[27] down, 2/9-tone down",
-      "35 medium diesis up, (35M, ~13M, ~125M, 5C\u00a0plus\u00a07C), 2/9-tone up",
+      QT_TRANSLATE_NOOP("symUserNames", "25 small diesis down, 2\u00b0 down [53 EDO]"),
+      QT_TRANSLATE_NOOP("symUserNames", "25 small diesis up, (25S, ~5:13S, ~37S, 5C\u00a0plus\u00a05C), 2\u00b0 up [53 EDO]"),
```
Comment (Jojo-Schmitz) [inline libmscore/sym.cpp:2711]: same here

File: libmscore/sym.cpp [hunk 4]
```diff
@@ -2707,12 +2707,12 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
...
+      QT_TRANSLATE_NOOP("symUserNames", "35 large diesis down, 2\u00b0 down [50 EDO], 5/18-tone down"),
+      QT_TRANSLATE_NOOP("symUserNames", "35 large diesis up, (35L, ~13L, ~125L,  sharp less 35M), 2\u00b050 up"),
```
Comment (Jojo-Schmitz) [inline libmscore/sym.cpp:2713]: and here

File: libmscore/sym.cpp [hunk 5]
```diff
@@ -2707,12 +2707,12 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
...
+      QT_TRANSLATE_NOOP("symUserNames", "35 medium diesis down, 1\u00b0[50] 2\u00b0[27] down, 2/9-tone down"),
+      QT_TRANSLATE_NOOP("symUserNames", "35 medium diesis up, (35M, ~13M, ~125M, 5C\u00a0plus\u00a07C), 2/9-tone up"),
```
Comment (Jojo-Schmitz) [inline libmscore/sym.cpp:2715]: and here

File: libmscore/sym.cpp [hunk 6]
```diff
@@ -2735,10 +2735,10 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
       "5:23 small diesis up, (5:23S, 5C plus 23C), 2\u00b0 up [60 EDO], 1/5-tone up",
       "5:49 medium diesis down",
       "5:49 medium diesis up, (5:49M, half apotome)",
-      "5:7 kleisma down",
-      "5:7 kleisma up, (5:7k, ~11:13k, 7C\u00a0less\u00a05C)",
-      "7 comma down, 1\u00b0 down [43 EDO], 2\u00b0 down [72 EDO], 1/6-tone down",
-      "7 comma up, (7C), 1\u00b0 up [43 EDO], 2\u00b0 up [72 EDO], 1/6-tone up",
+      QT_TRANSLATE_NOOP("symUserNames", "5:7 kleisma down"),
+      QT_TRANSLATE_NOOP("symUserNames", "5:7 kleisma up, (5:7k, ~11:13k, 7C\u00a0less\u00a05C)"),
```
Comment (Jojo-Schmitz) [inline libmscore/sym.cpp:2739]: and here

File: libmscore/types.h (hunk 1 above)
Comment (Ra-Yo-Su) [reply libmscore/types.h:314]: Would these work?:


SAGITTAL_5V7KU
--
SAGITTAL_5V7KD
SAGITTAL_5CU
SAGITTAL_5CD
SAGITTAL_7CU
SAGITTAL_7CD
SAGITTAL_25SDU
SAGITTAL_25SDD
SAGITTAL_35MDU
SAGITTAL_35MDD
SAGITTAL_11MDU
SAGITTAL_11MDD
SAGITTAL_11LDU
SAGITTAL_11LDD
SAGITTAL_35LDU
SAGITTAL_35LDD
SAGITTAL_SHARP25SD
SAGITTAL_FLAT25SU
SAGITTAL_SHARP7CD
SAGITTAL_FLAT7CU
SAGITTAL_SHARP5CD
SAGITTAL_FLAT5CU
SAGITTAL_SHARP5V7KD
SAGITTAL_FLAT5V7KU
SAGITTAL_SHARP
SAGITTAL_FLAT

Comment (Jojo-Schmitz) [reply libmscore/types.h:314]: yep, thanks.

**Comment by RobFog** [comment]
Rebase needed.

**Comment by Jojo-Schmitz** [comment]
Thanks, done

File: importexport/musicxml/musicxmlsupport.cpp [hunk 7]
```diff
@@ -375,74 +375,115 @@ QString accSymId2MxmlString(const SymId id)
...
+            case SymId::accidentalQuarterToneFlatNaturalArrowDown: s = "natural-down";         break;
+            case SymId::accidentalQuarterToneSharpNaturalArrowUp:  s = "natural-up";           break;
+            case SymId::accidentalThreeQuarterTonesFlatArrowDown:  s = "flat-down";            break;
+            case SymId::accidentalQuarterToneFlatArrowUp:          s = "flat-up";              break;
+
+            case SymId::accidentalArrowDown:             s = "arrow-down";           break;
+            case SymId::accidentalArrowUp:               s = "arrow-up";             break;
+
+            //case SymId::accidentalTripleSharp:           s = "triple-sharp";         break;
+            //case SymId::accidentalTripleFlat:            s = "triple-flat";          break;
```
Comment (Jojo-Schmitz) [inline importexport/musicxml/musicxmlsupport.cpp:404]: enable once #6002 is merged

File: importexport/musicxml/musicxmlsupport.cpp [hunk 8]
```diff
@@ -375,74 +375,115 @@ QString accSymId2MxmlString(const SymId id)
...
+      map["double-sharp-down"] = SymId::accidentalThreeQuarterTonesSharpArrowDown;
+      map["double-sharp-up"] = SymId::accidentalFiveQuarterTonesSharpArrowUp;
+      map["flat-flat-down"] = SymId::accidentalFiveQuarterTonesFlatArrowDown;
+      map["flat-flat-up"] = SymId::accidentalThreeQuarterTonesFlatArrowUp;
+
+      map["arrow-down"] = SymId::accidentalArrowDown;
+      map["arrow-up"] = SymId::accidentalArrowUp;
+
+      //map["triple-sharp"] = SymId::accidentalTripleSharp;
+      //map["triple-flat"] = SymId::accidentalTripleFlat;
```
Comment (Jojo-Schmitz) [inline importexport/musicxml/musicxmlsupport.cpp:467]: enable once #6002 is merged

File: importexport/musicxml/musicxmlsupport.cpp [hunk 9]
```diff
@@ -462,28 +503,55 @@ QString accidentalType2MxmlString(const AccidentalType type)
...
+            case AccidentalType::SHARP2_ARROW_DOWN:  s = "double-sharp-down";    break;
+            case AccidentalType::SHARP2_ARROW_UP:    s = "double-sharp-up";      break;
+            case AccidentalType::FLAT2_ARROW_DOWN:   s = "flat-flat-down";       break;
+            case AccidentalType::FLAT2_ARROW_UP:     s = "flat-flat-up";         break;
+
+            case AccidentalType::ARROW_DOWN:         s = "arrow-down";           break;
+            case AccidentalType::ARROW_UP:           s = "arrow-up";             break;
+
+            //case AccidentalType::SHARP3:             s = "triple-sharp";         break;
+            //case AccidentalType::FLAT3:              s = "triple-flat";          break;
```
Comment (Jojo-Schmitz) [inline importexport/musicxml/musicxmlsupport.cpp:535]: enable once #6002 is merged

File: importexport/musicxml/musicxmlsupport.cpp [hunk 10]
```diff
@@ -495,47 +563,67 @@ QString accidentalType2MxmlString(const AccidentalType type)
...
+      map["double-sharp-down"] = AccidentalType::SHARP2_ARROW_DOWN;
+      map["double-sharp-up"] = AccidentalType::SHARP2_ARROW_UP;
+      map["flat-flat-down"] = AccidentalType::FLAT2_ARROW_DOWN;
+      map["flat-flat-up"] = AccidentalType::FLAT2_ARROW_UP;
+
+      map["arrow-down"] = AccidentalType::ARROW_DOWN;
+      map["arrow-up"] = AccidentalType::ARROW_UP;
+
+      //map["triple-sharp"] = AccidentalType::SHARP3;
+      //map["triple-flat"] = AccidentalType::FLAT3;
```
Comment (Jojo-Schmitz) [inline importexport/musicxml/musicxmlsupport.cpp:602]: enable once #6002 is merged

File: importexport/musicxml/musicxmlsupport.cpp [hunk 11]
```diff
@@ -375,74 +375,115 @@ QString accSymId2MxmlString(const SymId id)
...
+      map["sharp-down"] = SymId::accidentalQuarterToneSharpArrowDown;
+      map["sharp-up"] = SymId::accidentalThreeQuarterTonesSharpArrowUp;
+      map["natural-down"] = SymId::accidentalQuarterToneFlatNaturalArrowDown;
+      map["natural-up"] = SymId::accidentalQuarterToneSharpNaturalArrowUp;
+      map["flat-down"] = SymId::accidentalThreeQuarterTonesFlatArrowDown;
+      map["flat-up"] = SymId::accidentalQuarterToneFlatArrowUp;
+      map["double-sharp-down"] = SymId::accidentalThreeQuarterTonesSharpArrowDown;
+      map["double-sharp-up"] = SymId::accidentalFiveQuarterTonesSharpArrowUp;
+      map["flat-flat-down"] = SymId::accidentalFiveQuarterTonesFlatArrowDown;
+      map["flat-flat-up"] = SymId::accidentalThreeQuarterTonesFlatArrowUp;
```
Comment (Jojo-Schmitz) [inline importexport/musicxml/musicxmlsupport.cpp:461]: See also #6263

File: importexport/musicxml/musicxmlsupport.cpp [hunk 12]
```diff
@@ -462,28 +503,55 @@ QString accidentalType2MxmlString(const AccidentalType type)
...
-            case AccidentalType::SHARP_ARROW_UP:     s = "sharp-up";             break;
+            case AccidentalType::SHARP_ARROW_DOWN:   s = "sharp-down";           break;
+            case AccidentalType::NATURAL_ARROW_UP:   s = "natural-up";           break;
+            case AccidentalType::NATURAL_ARROW_DOWN: s = "natural-down";         break;
+            case AccidentalType::FLAT_ARROW_DOWN:    s = "flat-down";            break;
+            case AccidentalType::FLAT_ARROW_UP:      s = "flat-up";              break;
+            case AccidentalType::SHARP2_ARROW_DOWN:  s = "double-sharp-down";    break;
+            case AccidentalType::SHARP2_ARROW_UP:    s = "double-sharp-up";      break;
+            case AccidentalType::FLAT2_ARROW_DOWN:   s = "flat-flat-down";       break;
+            case AccidentalType::FLAT2_ARROW_UP:     s = "flat-flat-up";         break;
```
Comment (Jojo-Schmitz) [inline importexport/musicxml/musicxmlsupport.cpp:529]: #6263

File: importexport/musicxml/musicxmlsupport.cpp [hunk 13]
```diff
@@ -495,47 +563,67 @@ QString accidentalType2MxmlString(const AccidentalType type)
...
       map["sharp-up"] = AccidentalType::SHARP_ARROW_UP;
       map["natural-down"] = AccidentalType::NATURAL_ARROW_DOWN;
       map["natural-up"] = AccidentalType::NATURAL_ARROW_UP;
+      map["sharp-down"] = AccidentalType::SHARP_ARROW_DOWN;
       map["flat-down"] = AccidentalType::FLAT_ARROW_DOWN;
       map["flat-up"] = AccidentalType::FLAT_ARROW_UP;
+      map["double-sharp-down"] = AccidentalType::SHARP2_ARROW_DOWN;
+      map["double-sharp-up"] = AccidentalType::SHARP2_ARROW_UP;
+      map["flat-flat-down"] = AccidentalType::FLAT2_ARROW_DOWN;
+      map["flat-flat-up"] = AccidentalType::FLAT2_ARROW_UP;
```
Comment (Jojo-Schmitz) [inline importexport/musicxml/musicxmlsupport.cpp:596]: See also #6263 (merge conflict waiting to happen;-))

File: importexport/musicxml/musicxmlsupport.cpp [hunk 14]
```diff
@@ -375,74 +375,119 @@ QString accSymId2MxmlString(const SymId id)
...
+            case SymId::accidentalQuarterToneSharpArrowDown:       s = "sharp-down";           break;
+            case SymId::accidentalThreeQuarterTonesSharpArrowUp:   s = "sharp-up";             break;
+            case SymId::accidentalQuarterToneFlatNaturalArrowDown: s = "natural-down";         break;
+            case SymId::accidentalQuarterToneSharpNaturalArrowUp:  s = "natural-up";           break;
+            case SymId::accidentalThreeQuarterTonesFlatArrowDown:  s = "flat-down";            break;
+            case SymId::accidentalQuarterToneFlatArrowUp:          s = "flat-up";              break;
+            case SymId::accidentalThreeQuarterTonesSharpArrowDown: s = "double-sharp-down";    break;
+            case SymId::accidentalFiveQuarterTonesSharpArrowUp:    s = "double-sharp-up";      break;
+            case SymId::accidentalFiveQuarterTonesFlatArrowDown:   s = "flat-flat-down";       break;
+            case SymId::accidentalThreeQuarterTonesFlatArrowUp:    s = "flat-flat-up";         break;
```
Comment (Jojo-Schmitz) [inline importexport/musicxml/musicxmlsupport.cpp:402]: See also #6263 (missing there)

File: libmscore/mscore.h [hunk 15]
```diff
@@ -213,7 +213,7 @@ enum class IconType : signed char {
       SBEAM, MBEAM, NBEAM, BEAM32, BEAM64, AUTOBEAM,
       FBEAM1, FBEAM2,
       VFRAME, HFRAME, TFRAME, FFRAME, MEASURE,
-      BRACKETS, PARENTHESES
+      BRACKETS, PARENTHESES, BRACES,
```
Comment (anatoly-os) [inline libmscore/mscore.h:216]: The last comma shouldn't be here

File: libmscore/accidental.cpp [hunk 16]
```diff
@@ -369,7 +439,7 @@ void Accidental::draw(QPainter* painter) const
             return;
       painter->setPen(curColor());
       for (const SymElement& e : el)
-            score()->scoreFont()->draw(e.sym, painter, magS(), QPointF(e.x, 0.0));
+            score()->scoreFont()->draw(e.sym, painter, magS(), QPointF(e.x, e.y));
```
Comment (anatoly-os) [inline libmscore/accidental.cpp:442]: Why? Looks like a serious change to the logic of drawing accidentals.

File: libmscore/sym.cpp [hunk 17]
```diff
@@ -2683,10 +2683,12 @@ const std::array<const char*, int(SymId::lastSym)+1> Sym::symUserNames = { {
       QT_TRANSLATE_NOOP("symUserNames", "No symbol"),
       "4-string tab clef",
       "6-string tab clef",
-      "11 large diesis down, 3\u00b0 down [46 EDO]",
-      "11 large diesis up, (11L), (sharp less 11M), 3\u00b0 up [46 EDO]",
-      "11 medium diesis down, 1\u00b0[17 31] 2\u00b046 down, 1/4-tone down",
-      "11 medium diesis up, (11M), 1\u00b0[17 31] 2\u00b046 up, 1/4-tone up",
+      QT_TRANSLATE_NOOP("symUserNames", "11 large diesis down, 3\u00b0 down [46 EDO]"),
```
Comment (anatoly-os) [inline libmscore/sym.cpp:2686]: Changing translatable strings? Not for 3.5.1 then, but up to 3.6.

File: libmscore/mscore.h (hunk 15 above)
Comment (Jojo-Schmitz) [reply libmscore/mscore.h:216]: Doesn't harm either and leads to leaner diffs on nedt extension

File: libmscore/sym.cpp (hunk 17 above)
Comment (Jojo-Schmitz) [reply libmscore/sym.cpp:2686]: The chord playback settings PRs (just merged) already add strings too

Comment (Jojo-Schmitz) [reply libmscore/sym.cpp:2686]: But 3.6 is fine by me, although that promise had been broken before already...

File: libmscore/accidental.cpp (hunk 16 above)
Comment (Jojo-Schmitz) [reply libmscore/accidental.cpp:442]: I don't recall, but omitting the y coordinates seems just plain wrong

File: libmscore/mscore.h (hunk 15 above)
Comment (Jojo-Schmitz) [reply libmscore/mscore.h:216]: See abobe, https://github.com/musescore/MuseScore/pull/6182/files#diff-f0e7e3ffeb78101e7c88d9bb470413bcR50

File: libmscore/sym.cpp (hunk 17 above)
Comment (Jojo-Schmitz) [reply libmscore/sym.cpp:2686]: Them not getting translated in time is not a big deal either

File: libmscore/accidental.cpp (hunk 16 above)
Comment (Jojo-Schmitz) [reply libmscore/accidental.cpp:442]: I believe it is due to the brackets, see https://github.com/musescore/MuseScore/pull/6182/files#diff-ecd72d8d5148bd7cab37cef6908b7d83R379 and https://github.com/musescore/MuseScore/pull/6182/files#diff-ecd72d8d5148bd7cab37cef6908b7d83R386

File: libmscore/sym.cpp (hunk 17 above)
Comment (Jojo-Schmitz) [reply libmscore/sym.cpp:2686]: You added it to the 3.5.1 milestones ;-) https://github.com/musescore/MuseScore/pull/6182#event-3470046745

File: libmscore/accidental.cpp [hunk 18]
```diff
@@ -362,23 +362,49 @@ void Accidental::layout()
...
+                        id = SymId::accidentalCombiningCloseCurlyBrace;
+                        break;
+                  case AccidentalBracket::NONE: // can't happen
+                        break;
+                  }
             x = r.x()+r.width();
-            SymElement e1(id, x);
-            el.append(e1);
+            SymElement e(id, x, _bracket == AccidentalBracket::BRACE ? spatium() * 0.4 : 0.0);
+            el.append(e);
```
Comment (Jojo-Schmitz) [inline libmscore/accidental.cpp:407]: Introduces a compiler warning, will fix it shortly

**Pull request merged**